espnet2/enh/layers/complexnn.py

```diff
@@ -0,0 +1,493 @@
+            )
+        else:
+            self.projection_dim = None
+
+    def forward(self, inputs):
+        if isinstance(inputs, list):
+            real, imag = inputs
+        elif isinstance(inputs, torch.Tensor):
+            real, imag = torch.chunk(inputs, -1)
+        r2r_out = self.real_lstm(real)[0]
+        r2i_out = self.imag_lstm(real)[0]
+        i2r_out = self.real_lstm(imag)[0]
+        i2i_out = self.imag_lstm(imag)[0]
+        real_out = r2r_out - i2i_out
+        imag_out = i2r_out + r2i_out
+        if self.projection_dim is not None:
+            real_out = self.r_trans(real_out)
+            imag_out = self.i_trans(imag_out)
+        # print(real_out.shape,imag_out.shape)
+        return [real_out, imag_out]
+
+    def flatten_parameters(self):
+        self.imag_lstm.flatten_parameters()
+        self.real_lstm.flatten_parameters()
+
+
+def complex_cat(inputs, axis):
+
+    real, imag = [], []
+    for idx, data in enumerate(inputs):
+        r, i = torch.chunk(data, 2, axis)
+        real.append(r)
+        imag.append(i)
+    real = torch.cat(real, axis)
+    imag = torch.cat(imag, axis)
+    outputs = torch.cat([real, imag], axis)
+    return outputs
+
+
+class ComplexConv2d(nn.Module):
+    def __init__(
+        self,
+        in_channels,
+        out_channels,
+        kernel_size=(1, 1),
+        stride=(1, 1),
+        padding=(0, 0),
+        dilation=1,
+        groups=1,
+        causal=True,
+        complex_axis=1,
+    ):
+        """
+            in_channels: real+imag
+            out_channels: real+imag
+            kernel_size : input [B,C,D,T] kernel size in [D,T]
+            padding : input [B,C,D,T] padding in [D,T]
+            causal: if causal, will padding time dimension's left side,
+                    otherwise both
+        """
+        super(ComplexConv2d, self).__init__()
+        self.in_channels = in_channels // 2
+        self.out_channels = out_channels // 2
+        self.kernel_size = kernel_size
+        self.stride = stride
+        self.padding = padding
+        self.causal = causal
+        self.groups = groups
+        self.dilation = dilation
+        self.complex_axis = complex_axis
+        self.real_conv = nn.Conv2d(
+            self.in_channels,
+            self.out_channels,
+            kernel_size,
+            self.stride,
+            padding=[self.padding[0], 0],
+            dilation=self.dilation,
+            groups=self.groups,
+        )
+        self.imag_conv = nn.Conv2d(
+            self.in_channels,
+            self.out_channels,
+            kernel_size,
+            self.stride,
+            padding=[self.padding[0], 0],
+            dilation=self.dilation,
+            groups=self.groups,
+        )
+
+        nn.init.normal_(self.real_conv.weight.data, std=0.05)
+        nn.init.normal_(self.imag_conv.weight.data, std=0.05)
+        nn.init.constant_(self.real_conv.bias, 0.0)
+        nn.init.constant_(self.imag_conv.bias, 0.0)
```

Comment:
We should be careful about the internal parameter initialization. It will be overridden by the global initialization in [espnet2/tasks/enh.py#L246-L247](https://github.com/espnet/espnet/blob/master/espnet2/tasks/enh.py#L246-L247).
Maybe it's better to add a line in the model config (egs2/dns_ins20/enh1/conf/tuning/train_enh_dccrn.yaml) to explicitly warn the users about it:
```
init: null   # do not set init method here because DCCRN has its own initialization
```

Reply:
OK